File: 279.完全平方数.cpp
/*
 * @lc app=leetcode.cn id=279 lang=cpp
 *
 * [279] 完全平方数
 */

// @lc code=start
class Solution {
public:
    int numSquares(int n) {
        vector<int> ps;
        for (int i = 1; i <= 100; i++) ps.push_back(i*i);
        int dp[n+1];
        dp[0] = 0;
        for (int i = 1; i <= n; i++) {
            dp[i] = i;
            for (int j = 0; j < ps.size() && ps[j] <= i; j++) {
                dp[i] = min(dp[i], dp[i-ps[j]] + 1);
            }
        }
        return dp[n];
    }
};
// @lc code=end

File: 279.完全平方数.py
#
# @lc app=leetcode.cn id=279 lang=python3
#
# [279] 完全平方数
#

# @lc code=start
class Solution:
    def numSquares(self, n: int) -> int:
        squares = []
        i = 0
        while i*i <= n:
            squares.append(i*i)
            i += 1
        dp = [0] * (n+1)
        dp[0] = 0
        for j in range(1, n+1):
            dp[j] = j
            i = 0
            while i < len(squares) and squares[i] <= j:
                dp[j] = min(dp[j], dp[j-squares[i]]+1)
                i += 1
        return dp[n]
# @lc code=end

